### PR TITLE
add FreeBSD support

### PR DESCRIPTION
**Pull request by mmoll**
Opening this mainly for comments, tests are not yet adjusted.
I'm unsure if it's better to have all the information per osfamily in params.pp or only the differing variables but did it the first way as this is easier for me at the moment.

### CODE REVIEW

File: manifests/config.pp [hunk 1]
```diff
@@ -24,7 +24,7 @@
 
   user { $foreman_proxy::user:
     ensure  => 'present',
-    shell   => '/bin/false',
+    shell   => $::foreman_proxy::params::false_bin,
```
Comment (ekohl) [inline manifests/config.pp]: Why not name it ```$shell``` or similar?

File: manifests/config.pp [hunk 2]
```diff
@@ -73,22 +73,22 @@
...
-      file { '/etc/sudoers.d/foreman-proxy':
+      file { "${::foreman_proxy::sudoers}.d/foreman-proxy":
         ensure  => file,
         owner   => 'root',
-        group   => 'root',
+        group   => 0,
         mode    => '0440',
         content => template('foreman_proxy/sudo.erb'),
-        require => File['/etc/sudoers.d'],
+        require => File["${::foreman_proxy::sudoers}.d"],
```
Comment (ekohl) [inline manifests/config.pp]: This is redundant since puppet 2.7

**Comment by ekohl** [comment]
I think params.pp is too duplicated now.

**Comment by mmoll** [comment]
Another round of review, please. This time _only_ FreeBSD tests should fail.

File: manifests/config.pp [hunk 3]
```diff
@@ -24,7 +24,7 @@
 
   user { $foreman_proxy::user:
     ensure  => 'present',
-    shell   => '/bin/false',
+    shell   => $::foreman_proxy::params::shell,
```
Comment (ekohl) [inline manifests/config.pp]: Since init inherits from params, the ```::params``` is not needed and would allow for easier overriding in the future.

**Comment by ekohl** [comment]
Would a variable ```$etc``` (or something similar) in params work against the redundancy? Even having a variable for ```/etc/foreman-proxy``` or ```/usr/local/etc/foreman-proxy``` would make it much shorter.

**Comment by mmoll** [comment]
something like this?
```puppet
case $::osfamily {
  /^(FreeBSD|DragonFly)$/: {
    $etc = '/usr/local/etc'
  }
  default: {
    $etc = '/etc'
  }
}
```

**Comment by ekohl** [comment]
That was what I was thinking about. No idea how readable it will be when you do use that.

**Comment by mmoll** [comment]
I think it is a tad better.

**Comment by mmoll** [comment]
@ekohl @domcleal if this is OK for you code-wise as-is, I'd go on with adding FreeBSD tests.

**Comment by domcleal** [comment]
Yeah, fine I think.  I'm not sure about rearranging params into large per-OS blocks as it splits up where parameters for each module are defined, but otherwise OK.

**Comment by mmoll** [comment]
This is ready now. There's still one failing test I don't know how to tackle, probbaly caused by the usage of `ensure_packages` and `bind910` already being installed by the dns class.

**Comment by mmoll** [comment]
The other test failures are caused by rspec-puppet-facts 1.1.0 now setting the `puppetversion` fact correctly and our tests don't take the changed params from puppet-puppet according to the puppet version into account.

**Comment by mmoll** [comment]
Rebased. This leaves only one test error and I have no idea how to properly test `ensure_packages` in such a situation.

**Comment by domcleal** [comment]
> I have no idea how to properly test ensure_packages in such a situation.

I think the problem is that puppet-dns doesn't use ensure_packages, so it's a little unpredictable as to what name the package resource gets.  I'm surprised that it passes for some OSes and not others.

The catalog in the test contain `Package[dns]{:name=>"bind910", :ensure=>"present"}` which is from https://github.com/theforeman/puppet-dns/blob/master/manifests/install.pp rather than ensure_packages, and this is why the expectation in this module's not matching.

I think if puppet-dns used ensure_packages then at least the name would be consistently OS-specific and we could test for that in both modules.

**Comment by mmoll** [comment]
:green_heart:

File: manifests/params.pp [hunk 4]
```diff
@@ -58,6 +64,33 @@
                                     '/usr/lib/syslinux/pxelinux.0']
       }
     }
+    /^(FreeBSD|DragonFly)$/: {
+      # if set to true, no repo will be added by this module, letting you to
+      # set it to some custom location.
+      $custom_repo         = true # as foreman packages are in standard FreeBSD ports
+      $plugin_prefix       = 'rubygem-smart_proxy_'
+      $foreman_api_package = 'rubygem-apipie-bindings'
```
Comment (domcleal) [inline manifests/params.pp]: Are these right?

**Comment by domcleal** [comment]
:+1: aside from one query.

File: manifests/params.pp (hunk 4 above)
Comment (mmoll) [inline manifests/params.pp]: Yes, see http://portscout.freebsd.org/mmoll@freebsd.org.html, DragonFly uses the same ports tree, so everything is named like on FreeBSD.

**Comment by ekohl** [comment]
Can't really speak for BSD, but :+1: for the code and I trust you on the BSD part.

**Comment by domcleal** [comment]
Thanks @mmoll!

**Comment by mmoll** [comment]
woohoo :smiling_imp:

**Pull request merged**